File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        int n = s.size();

        vector<vector<int>>dp(n+1, vector<int>(n+1, 0));
        int startIndex = 0;
        int maxLen = 0;
        
        for(int diff = 0; diff < n; diff++) {
            for(int i=0, j=diff; j<n; i++, j++) {
                if(diff == 0) {
                    dp[i][j] = 1;
                }
                else if(diff == 1 && s[i] == s[j]) {
                   dp[i][j] = 2;
                }
                else {
                    if(s[i] == s[j] && dp[i+1][j-1]) {
                        dp[i][j] = dp[i+1][j-1] + 2;
                    }

                }
                if(maxLen < dp[i][j]) { 
                    maxLen = dp[i][j];
                    startIndex = i;
                }
            }
        }
        return s.substr(startIndex, maxLen);
    }
};

File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.py
class Solution(object):
    def longestPalindrome(self, s):
        """
        :type s: str
        :rtype: str
        """

        startIndex = -1
        maxLen = 0

        for i in range(0, len(s)):
            low = i
            high = i

            while(low >= 0 and high < len(s)):
                if s[low] == s[high]:
                    if maxLen < high-low+1:
                        startIndex = low
                        maxLen = high-low+1
                    low -= 1
                    high += 1
                else:
                    break
        
        for i in range(0, len(s)-1):
            low = i
            high = i+1

            while(low >= 0 and high < len(s)):
                if s[low] == s[high]:
                    if maxLen < high-low+1:
                        startIndex = low
                        maxLen = high-low+1
                    low = low - 1
                    high = high + 1
                else:
                    break
        return s[startIndex:startIndex+maxLen]